SongSnap_FrozenWeights.ipynb

```diff
@@ -0,0 +1,3550 @@
+        "        total_recall_genre += recall_score(preds_genre.cpu(), batch_y_genre.long().cpu(), average = 'macro')\n",
+        "        total_f1_genre += f1_score(preds_genre.cpu(), batch_y_genre.long().cpu(), average = 'macro')\n",
+        "\n",
+        "        _, preds_pop = torch.max(popularity_output, 1)\n",
+        "        total_accuracy_pop += accuracy_score(preds_pop.cpu(), batch_y_popularity.cpu())\n",
+        "        total_precision_pop += precision_score(preds_pop.cpu(), batch_y_popularity.cpu(), average = 'macro')\n",
+        "        total_recall_pop += recall_score(preds_pop.cpu(), batch_y_popularity.cpu(), average = 'macro')\n",
+        "        total_f1_pop += f1_score(preds_pop.cpu(), batch_y_popularity.cpu(), average = 'macro')\n",
+        "\n",
+        "        # Save the outputs to lists\n",
+        "        all_genre_outputs.extend(preds_genre.cpu().numpy())\n",
+        "        all_popularity_outputs.extend(preds_pop.cpu().numpy())\n",
+        "        all_danceability_outputs.extend(danceability_output.cpu().numpy())\n",
+        "        all_energy_outputs.extend(energy_output.cpu().numpy())\n",
+        "\n",
+        "        num_batches += 1\n",
+        "\n",
+        "    # Calculate average loss\n",
+        "    avg_loss_genre = total_loss_genre / num_batches\n",
+        "    avg_loss_popularity = total_loss_popularity / num_batches\n",
+        "    avg_loss_danceability = total_loss_danceability / num_batches\n",
+        "    avg_loss_energy = total_loss_energy / num_batches\n",
+        "\n",
+        "    avg_acc_genre = total_accuracy_genre / num_batches\n",
+        "    avg_precision_genre = total_precision_genre / num_batches\n",
+        "    avg_recall_genre = total_recall_genre / num_batches\n",
+        "    avg_f1_genre = total_f1_genre / num_batches\n",
+        "\n",
+        "    avg_acc_pop = total_accuracy_pop / num_batches\n",
+        "    avg_precision_pop = total_precision_pop / num_batches\n",
+        "    avg_recall_pop = total_recall_pop / num_batches\n",
+        "    avg_f1_pop = total_f1_pop / num_batches\n",
+        "\n",
+        "    # Calculate the average accuracy, recall, precision and F1 score for genre & popularity\n",
+        "\n",
+        "\n",
+        "    print(\"Evaluation Results:\")\n",
+        "    print(\"Average Genre loss: {:.4f}\".format(avg_loss_genre))\n",
+        "    print(\"Average Popularity loss: {:.4f}\".format(avg_loss_popularity))\n",
+        "    print(\"Average Danceability loss: {:.4f}\".format(avg_loss_danceability))\n",
+        "    print(\"Average Energy loss: {:.4f}\".format(avg_loss_energy))\n",
+        "\n",
+        "    print(\" \")\n",
+        "    print(\"Average Accuracy Popularity: {:.4f}\".format(avg_acc_pop))\n",
+        "    print(\"Average Precision Popularity: {:.4f}\".format(avg_precision_pop))\n",
```

Comment:
Good idea to record all the metrics separately and individually for each target.